File: languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Envelope/KeyMetaTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.Envelope;
using GoDaddy.Asherah.Crypto.ExtensionMethods;
using Newtonsoft.Json.Linq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Envelope
{
    [Collection("Logger Fixture collection")]
    public class KeyMetaTest
    {
        private const string KeyId = "key1";
        private readonly DateTimeOffset created = DateTimeOffset.UtcNow.Truncate(TimeSpan.FromMinutes(1));

        private readonly KeyMeta keyMeta;

        public KeyMetaTest()
        {
            keyMeta = new KeyMeta(KeyId, created);
        }

        [Fact]
        private void TestConstructorRegular()
        {
            KeyMeta meta = new KeyMeta(KeyId, created);
            Assert.NotNull(meta);
            Assert.Equal(KeyId, meta.KeyId);
            Assert.Equal(created, meta.Created);
        }

        [Fact]
        private void TestConstructorJson()
        {
            Asherah.AppEncryption.Util.Json keyMetaJson = new Asherah.AppEncryption.Util.Json();
            keyMetaJson.Put("KeyId", KeyId);
            keyMetaJson.Put("Created", created);

            KeyMeta meta = new KeyMeta(keyMetaJson);
            Assert.NotNull(meta);
            Assert.Equal(KeyId, meta.KeyId);
            Assert.Equal(created, meta.Created);
        }

        [Fact]
        private void TestToJson()
        {
            JObject metaJson = keyMeta.ToJson();
            Assert.Equal(KeyId, metaJson.GetValue("KeyId").ToObject<string>());
            Assert.Equal(created.ToUnixTimeSeconds(), metaJson.GetValue("Created").ToObject<long>());
        }

        [Fact]
        private void TestHashCodeAndEqualsSymmetric()
        {
            KeyMeta otherKeyMeta = new KeyMeta(KeyId, created);
            Assert.True(keyMeta.GetHashCode() == otherKeyMeta.GetHashCode());
            Assert.NotSame(keyMeta, otherKeyMeta);
            Assert.True(keyMeta.Equals(otherKeyMeta) && otherKeyMeta.Equals(keyMeta));
        }

        [Fact]
        private void TestEqualsWithSameInstance()
        {
            // ReSharper disable once EqualExpressionComparison
            Assert.True(keyMeta.Equals(keyMeta));
        }

        [Fact]
        private void TestEqualsWithNull()
        {
            Assert.False(keyMeta.Equals(null));
        }

        [Fact]
        private void TestEqualsWithDifferentClass()
        {
            // ReSharper disable once SuspiciousTypeConversion.Global
            Assert.False(keyMeta.Equals("blah"));
        }

        [Fact]
        private void TestEqualsWithDifferentKeyId()
        {
            KeyMeta otherKeyMeta = new KeyMeta("some_other_keyid", created);
            Assert.False(keyMeta.Equals(otherKeyMeta));
            Assert.False(otherKeyMeta.Equals(keyMeta));
        }

        [Fact]
        private void TestEqualsWithDifferentCreated()
        {
            KeyMeta otherKeyMeta = new KeyMeta(KeyId, created.AddMinutes(3));
            Assert.False(keyMeta.Equals(otherKeyMeta));
            Assert.False(otherKeyMeta.Equals(keyMeta));
        }
    }
}

File: languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/envelope/KeyMetaTest.java
package com.godaddy.asherah.appencryption.envelope;

import static org.junit.jupiter.api.Assertions.*;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import org.json.JSONObject;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.godaddy.asherah.appencryption.utils.Json;

class KeyMetaTest {
  private final String keyId = "key1";
  private final Instant created = Instant.now().truncatedTo(ChronoUnit.SECONDS);

  private KeyMeta keyMeta;

  @BeforeEach
  void setUp(){
    keyMeta = new KeyMeta(keyId, created);
  }

  @Test
  void testConstructorRegular() {
    KeyMeta meta = new KeyMeta(keyId, created);
    assertNotNull(meta);
    assertEquals(keyId, meta.getKeyId());
    assertEquals(created, meta.getCreated());
  }

  @Test
  void testConstructorJson() {
    Json keyMetaJson = new Json();
    keyMetaJson.put("KeyId", keyId);
    keyMetaJson.put("Created", created);

    KeyMeta meta = new KeyMeta(keyMetaJson);
    assertNotNull(meta);
    assertEquals(keyId, meta.getKeyId());
    assertEquals(created, meta.getCreated());
  }

  @Test
  void testToJson() {
    JSONObject metaJson = keyMeta.toJson();
    assertEquals(keyId, metaJson.getString("KeyId"));
    assertEquals(created.getEpochSecond(), metaJson.getLong("Created"));
  }

  @Test
  void testHashCodeAndEqualsSymmetric() {
    KeyMeta otherKeyMeta = new KeyMeta(keyId, created);
    assertTrue(keyMeta.hashCode() == otherKeyMeta.hashCode());
    assertNotSame(keyMeta, otherKeyMeta);
    assertTrue(keyMeta.equals(otherKeyMeta) && otherKeyMeta.equals(keyMeta));
  }

  @Test
  void testEqualsWithSameInstance() {
    assertTrue(keyMeta.equals(keyMeta));
  }

  @Test
  void testEqualsWithNull() {
    assertFalse(keyMeta.equals(null));
  }

  @SuppressWarnings("unlikely-arg-type")
  @Test
  void testEqualsWithDifferentClass() {
    assertFalse(keyMeta.equals("blah"));
  }

  @Test
  void testEqualsWithDifferentKeyId() {
    KeyMeta otherKeyMeta = new KeyMeta("some_other_keyid", created);
    assertFalse(keyMeta.equals(otherKeyMeta));
    assertFalse(otherKeyMeta.equals(keyMeta));
  }

  @Test
  void testEqualsWithDifferentCreated() {
    KeyMeta otherKeyMeta = new KeyMeta(keyId, created.minus(1, ChronoUnit.MINUTES));
    assertFalse(keyMeta.equals(otherKeyMeta));
    assertFalse(otherKeyMeta.equals(keyMeta));
  }

}
